lib/modules/manager/nix/extract.spec.ts

```diff
@@ -1,76 +1,508 @@
+    expect((await extractPackageFile('', 'flake.nix'))?.deps).toEqual([
       {
         depName: 'nixpkgs',
-        currentValue: 'nixos-21.11',
+        currentDigest: '9f4128e00b0ae8ec65918efeba59db998750ead6',
+        currentValue: 'nixos-unstable',
         datasource: GitRefsDatasource.id,
         packageName: 'https://github.com/NixOS/nixpkgs',
-        versioning: nixpkgsVersioning,
+        replaceString: '9f4128e00b0ae8ec65918efeba59db998750ead6',
       },
     ]);
   });
 
-  it('is case insensitive', () => {
-    const content = `{
-  inputs = {
-    nixpkgs.url = "github:NixOS/nixpkgs/nixos-21.11";
-  };
-}`;
+  it('includes nixpkgs with no explicit ref', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(flake3Lock);
+    expect((await extractPackageFile('', 'flake.nix'))?.deps).toMatchObject([
+      {
+        currentDigest: '612ee628421ba2c1abca4c99684862f76cb3b089',
+        datasource: 'git-refs',
+        depName: 'nixpkgs',
+        packageName: 'https://github.com/NixOS/nixpkgs',
+      },
+    ]);
+  });
 
-    const res = extractPackageFile(content);
+  it('includes patchelf from HEAD', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(flake4Lock);
+    expect((await extractPackageFile('', 'flake.nix'))?.deps).toMatchObject([
+      {
+        currentDigest: 'a0f54334df36770b335c051e540ba40afcbf8378',
+        datasource: 'git-refs',
+        depName: 'patchelf',
+        packageName: 'https://github.com/NixOS/patchelf.git',
+      },
+    ]);
+  });
+
+  it('includes ijq from sourcehut without a flake', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(flake5Lock);
+    expect((await extractPackageFile('', 'flake.nix'))?.deps).toMatchObject([
+      {
+        currentDigest: '88f0d9ae98942bf49cba302c42b2a0f6e05f9b58',
+        datasource: 'git-refs',
+        depName: 'ijq',
+        packageName: 'https://git.sr.ht/~gpanders/ijq',
+      },
+    ]);
+  });
 
-    expect(res?.deps).toEqual([
+  it('includes home-manager from gitlab', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(flake6Lock);
+    expect((await extractPackageFile('', 'flake.nix'))?.deps).toMatchObject([
       {
+        currentDigest: '65ae9c147349829d3df0222151f53f79821c5134',
+        datasource: 'git-refs',
+        depName: 'home-manager',
+        packageName: 'https://gitlab.com/rycee/home-manager',
+      },
+    ]);
+  });
+
+  it('test other version', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(flake7Lock);
+    expect(await extractPackageFile('', 'flake.nix')).toBeNull();
+  });
+
+  it('includes nixpkgs with ref and shallow arguments', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(flake8Lock);
+    expect((await extractPackageFile('', 'flake.nix'))?.deps).toMatchObject([
+      {
+        currentDigest: '5633bcff0c6162b9e4b5f1264264611e950c8ec7',
+        datasource: 'git-refs',
         depName: 'nixpkgs',
-        currentValue: 'nixos-21.11',
```

Comment:
is this still supported to update in the flake file?

Reply:
Nope but I would argue that this is the least useful feature for the nix package manager. For major updates especially on NixOS people should read the changelog as not every change can be notified via a warning/error and it contains extra information for special cases. Also the showed warnings should be fixed as they are turned into errors the next release or might be removed.

Reply:
I think there's an argument to be made that updating `flake.nix` references to the latest nixpkgs release stream would be good behavior to have.  Of course, this type of update isn't expected to be a quick non-breaking change, but Renovate's scope isn't limited to just updating things that are non-breaking.

I think that the scope of doing this correctly is extremely narrow... which I guess is what was present before this PR. 🙂

As a user, if I had to choose between "getting all my flake inputs updated regularly" (this PR) versus "getting my nixpkgs release reference updated twice a year", I would definitely chose this PR.  The cases where I want to use Renovate are going to be references to `unstable` (development environments), not my desktop/server where I want stable packages.

But I suppose it's a fair question -- does it have to be a choice, and why both can't be?

Reply:
> does it have to be a choice, and why both can't be?

my typescript knowledge to develop that is not good enough 😅